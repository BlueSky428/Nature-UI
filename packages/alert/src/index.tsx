import * as React from 'react';
import clx from 'clsx';
import { FiInfo, FiAlertCircle, FiAlertTriangle, FiCheckCircle } from 'react-icons/fi';

const SUBTLE_TEXT = 'text-gray-800';

export const ALERT_STATUSES = {
  error: {
    bg: 'bg-red-200',
    iconColor: 'text-red-600 mr-3',
    icon: FiAlertCircle,
    variant: {
      solid: 'bg-red-600 text-white',
      subtle: SUBTLE_TEXT,
    },
  },
  info: {
    bg: 'bg-blue-200',
    iconColor: 'text-blue-600 mr-3',
    icon: FiInfo,
    variant: {
      solid: 'bg-blue-600',
      subtle: SUBTLE_TEXT,
    },
  },
  success: {
    bg: 'bg-green-200',
    iconColor: 'text-green-600 mr-3',
    icon: FiCheckCircle,
    variant: {
      solid: 'bg-green-600 text-white',
      subtle: SUBTLE_TEXT,
    },
  },
  warning: {
    bg: 'bg-orange-200',
    iconColor: 'text-orange-600 mr-3',
    icon: FiAlertTriangle,
    variant: {
      solid: 'bg-orange-600 text-white',
      subtle: SUBTLE_TEXT,
    },
  },
};

interface AlertOptions {
  /**
   * The status of the alert
   */
  status?: keyof typeof ALERT_STATUSES;
  /**
   * The variant of the alert style to use
   */
  variant?: 'solid' | 'subtle';

  /**
   * Other tailwind utility classes or custom classnames you wish to include
   */
  className?: string;

  children?: React.ReactNode;
  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   */
  component?: React.ElementType;
}

export type AlertProps = AlertOptions;

const BASE_STYLE = 'px-4 py-3 flex items-center';

const Alert: React.ForwardRefExoticComponent<AlertProps> = React.forwardRef(
  (props: AlertProps, ref: React.Ref<any>) => {
    const {
      className = '',
      status = 'success',
      children,
      variant = 'subtle',
      component: Component = 'div',
    } = props;

    // const Component = 'div';

    const VARIANT: string = ALERT_STATUSES[status].variant[variant];

    const componentClass = clx(BASE_STYLE, {
      [className]: className,
      [ALERT_STATUSES[status].bg]: status,
      [VARIANT]: variant,
    });

    const Icon = ALERT_STATUSES[status];
    const iconClasses = clx({
      [VARIANT]: variant,
      'mr-3': variant,
      [Icon.iconColor]: variant !== 'solid',
    });

    const IconComponent = Icon.icon;

    return (
      <Component className={componentClass} ref={ref}>
        {status && <IconComponent className={iconClasses} size={20} />}
        <span>{children ? children : 'This is an alert ðŸ™‚'}</span>
      </Component>
    );
  }
);

Alert.displayName = 'Alert';

export default Alert;
