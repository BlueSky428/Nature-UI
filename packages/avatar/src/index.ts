export * from './avatar';
export * from './avatarGroup';
