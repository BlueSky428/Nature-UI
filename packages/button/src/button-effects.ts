import { css, keyframes } from '@nature-ui/system';

export const _ripple = keyframes`
      50% {
        opacity: 0.3;
      }
      100% {
        opacity: 0;
        transform: translate(-50%, -50%) scale(10);
      }
      `;

export const rippleEffect = css`
  line-height: 1.2;

  &::after {
    content: '';
    position: absolute;
    top: 50%;
    left: 50%;
    width: 1em;
    height: 1em;
    background: currentColor;
    border-radius: 50%;
    opacity: 0;
  }

  &:focus:not(:active)::after {
    animation: 0.3s ${_ripple};
  }
`;
