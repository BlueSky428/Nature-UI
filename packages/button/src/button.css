.button--ripple::after {
  content: '';
  position: absolute;
  top: 50%;
  left: 50%;
  width: 1em;
  height: 1em;
  background: currentColor;
  border-radius: 50%;
  opacity: 0;
}
.button--ripple:focus:not(:active)::after {
  animation: 0.3s ripple;
}

@keyframes ripple {
  /* 0% {
    opacity: 0;
    transform: translate(-50%, -50%) scale(0);
  } */
  50% {
    opacity: 0.3;
  }
  100% {
    opacity: 0;
    transform: translate(-50%, -50%) scale(10);
  }
}
