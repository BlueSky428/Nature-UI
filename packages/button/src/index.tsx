import * as React from 'react';
import clx from 'clsx';

interface ButtonProps {
  /**
   * The label of the button
   */
  'aria-label'?: string;
  /**
   * The text color of the button. Use a color key passed in theme.colors.
   */
  text?: string;
  /**
   * The background color of the button. Use a color key passed in theme.colors.
   */
  color?: string;
  /**
   * The variant of the button style to use.
   */
  variant?: 'outline' | 'ghost' | 'link' | 'solid' | 'none';
  /**
   * If true, the button will be disabled.
   */
  isDisabled?: boolean;
  /**
   * If true, the button will show a spinner.
   */
  isLoading?: boolean;
  /**
   * The label to show in the button when isLoading is true. If no text is passed, it only shows the spinner
   */
  loadingText?: string;
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'none';
  /**
   * Other tailwind utility classes or custom classnames you wish to include
   */
  className?: string;
  /**
   * The lightness or how dark you want the background color to be
   */
  children?: React.ReactNode;
  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   * defaults to "button"
   */
  component?: React.ElementType;
  id?: string;
  /**
   *  when the href prop is set it changes the tag name to a
   */
  href?: string;
  onClick?: () => any;
}

const darken = (value: string): string => {
  const splitStr = value.split('-');
  const dept = Number(splitStr[1]);
  const color = splitStr[0];

  if (dept >= 800) {
    return `${color}-${dept - 200}`;
  }

  return `${color}-${dept + 200}`;
};

const lighten = (value: string): string => {
  const color = value.split('-')[0];

  return `${color}-${100}`;
};

const Button = React.forwardRef((props: ButtonProps, ref: React.Ref<any>) => {
  const {
    component: Component = 'button',
    variant = 'solid',
    color = 'teal-500',
    text = 'white',
    size = 'md',
    children,
    ...rest
  } = props;
  const DEFAULT_CLASS = 'block focus:shadow-outline focus:outline-none rounded font-semibold flex';
  const STYLES = {
    solid: `bg-${color} text-${text} hover:bg-${darken(color)}`,
    outline: `bg-transparent hover:bg-${lighten(
      text
    )} text-${text} border border-${text} focus:border-transparent`,
    ghost: `hover:bg-${lighten(text)} text-${text}`,
    link: `hover:underline text-${text}`,
    md: 'px-4 py-2',
  };

  let BTNClass: string;

  if (variant === 'none') {
    BTNClass = '';
  } else {
    BTNClass = clx(DEFAULT_CLASS, {
      [STYLES[size]]: size && variant !== 'link',
      [STYLES[variant]]: variant,
    });
  }

  return (
    <Component className={BTNClass} ref={ref} {...rest}>
      {children ? children : 'Button'}
    </Component>
  );
});

export default Button;
