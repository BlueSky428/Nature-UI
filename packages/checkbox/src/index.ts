export * from './checkbox';
export * from './checkbox-group';
export * from './checkbox-types';
export * from './use-checkbox';
export * from './use-checkbox-group';
