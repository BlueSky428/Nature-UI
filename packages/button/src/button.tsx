/** ** */
import { Spinner } from '@nature-ui/spinner';
import {
  clsx,
  css,
  forwardRef,
  HTMLNatureProps,
  nature,
} from '@nature-ui/system';
import {
  darken,
  dataAttr,
  lighten,
  StringOrNumber,
  __DEV__,
} from '@nature-ui/utils';
import * as React from 'react';
import { rippleEffect } from './button-effects';

interface ButtonOptions {
  /**
   * The text color of the button. Use a color key passed in theme.colors.
   */
  text?: string;
  /**
   * The background color of the button. Use a color key passed in theme.colors.
   */
  color?: string;
  /**
   * The variant of the button style to use.
   */
  variant?: 'outline' | 'ghost' | 'link' | 'solid' | 'none';
  /**
   * If true, the button will be disabled.
   */
  isDisabled?: boolean;
  /**
   * If `true`, the button will be styled in it's active state.
   */
  isActive?: boolean;
  /**
   * If true, the button will show a spinner.
   */
  isLoading?: boolean;
  /**
   * The label to show in the button when isLoading is true. If no text is passed, it only shows the spinner
   */
  loadingText?: string;
  size?: 'xs' | 'sm' | 'md' | 'lg' | number;
  /**
   * If added, the button will show an icon before the button's label.
   * @type React.ReactElement
   */
  leftIcon?: React.ReactElement;
  /**
   * If added, the button will show an icon after the button's label.
   * @type React.ReactElement
   */
  rightIcon?: React.ReactElement;
  /**
   * The space between the button icon and label.
   * @type SystemProps["marginRight"]
   */
  iconSpacing?: StringOrNumber;
  /**
   * Typeof String | JSX.Element
   */
  children?: React.ReactNode;
  /**
   * to specify if a button is an icon button.
   * If this is set to true, the default padding-x will be removed
   */
  isIconButton?: boolean;
}

export interface ButtonProps extends HTMLNatureProps<'button'>, ButtonOptions {}

const _SIZES = {
  xs: {
    size: '1.5rem',
    font: '0.75rem',
    padding: 'px-2',
  },
  sm: {
    size: '2rem',
    font: '0.875rem',
    padding: 'px-3',
  },
  md: {
    size: '2.5rem',
    font: '1rem',
    padding: 'px-4',
  },
  lg: {
    size: '3rem',
    font: '1.125rem',
    padding: 'px-5',
  },
};

const ButtonIcon = forwardRef<HTMLNatureProps<'span'>, 'span'>((props, ref) => {
  const { children, ...rest } = props;

  const _children = React.isValidElement(children)
    ? React.cloneElement(children, {
        'aria-hidden': true,
        focusable: false,
      })
    : children;

  return (
    <nature.span ref={ref} {...rest}>
      {_children}
    </nature.span>
  );
});

if (__DEV__) {
  ButtonIcon.displayName = 'ButtonIcon';
}

export const ButtonSpinner = (
  props: ButtonProps & {
    spinner?: React.ReactNode;
    label?: string;
  },
) => {
  const { className = '', label, spinner, ...rest } = props;

  const _className = clsx(
    'align-middle',
    {
      absolute: !label,
      relative: label,
    },
    className,
  );

  return (
    <span className={_className} {...rest}>
      {spinner || <Spinner size='xs' color='currentColor' />}
      {label && <span className='ml-2'>{label}</span>}
    </span>
  );
};

export const Button = forwardRef<ButtonProps, 'button'>((props, ref) => {
  const {
    variant = 'solid',
    color = 'gray-200',
    size = 'md',
    children,
    text: _text,
    className = '',
    isDisabled = false,
    disabled = false,
    isLoading = false,
    loadingText,
    isActive,
    leftIcon,
    rightIcon,
    isIconButton,
    iconSpacing = '10px',
    ...rest
  } = props;

  const textColor = (): string => {
    const split = color.split('-');
    const amount = Number(split[split.length - 1]);
    if (amount >= 300) {
      return 'white';
    }
    if (!amount) {
      return 'white';
    }
    return 'gray-600';
  };

  let text = _text || textColor();

  const _size = typeof size === 'string' ? _SIZES[size].size : `${size}px`;
  const _font = _SIZES[size].font ?? '1rem';
  const _padding = _SIZES[size].padding ?? '0.8rem';
  const _disabled = isDisabled || disabled || isLoading;

  const _link = variant === 'link';

  const _sizes = css({
    height: _size,
    minWidth: _size,
  });

  if (Number(color.split('-')[1]) <= 400) {
    text = 'gray-700';
  }

  if (variant === 'ghost') {
    text = _text || 'gray-700';
  }

  const DEFAULT_CLASS =
    'focus:outline-none rounded font-semibold relative overflow-hidden align-middle inline-flex justify-center items-center leading-normal';
  const STYLES = {
    solid: `bg-${color} text-${text} hover:bg-${darken(color)}`,
    outline: `bg-transparent text-${color ?? text} border border-${
      color ?? text
    } focus:border-transparent hover:bg-${lighten(color ?? text)}`,
    ghost: `hover:bg-${lighten(text)} text-${text}`,
    link: `hover:underline text-${color}`,
    disabled: 'opacity-50 cursor-not-allowed',
  };

  let BTNClass: string;

  if (variant === 'none') {
    BTNClass = clsx(
      {
        [STYLES.disabled]: _disabled,
      },
      className,
    );
  } else {
    BTNClass = clsx(
      rippleEffect,
      DEFAULT_CLASS,
      {
        [_sizes]: !_link,
        [STYLES[variant]]: variant,
        [STYLES.disabled]: _disabled,
        [_padding]: !isIconButton && !_link,
      },
      className,
    );
  }

  const defaults = {
    className: BTNClass,
    ref,
    size,
    disabled: _disabled,
    'data-active': dataAttr(isActive),
    'data-loading': dataAttr(isLoading),
  };

  return (
    <nature.button
      css={{
        fontSize: _font,
      }}
      {...defaults}
      {...rest}
    >
      {leftIcon && !isLoading && (
        <ButtonIcon
          css={{
            marginRight: iconSpacing,
          }}
        >
          {leftIcon}
        </ButtonIcon>
      )}
      {isLoading ? (
        <>
          <ButtonSpinner label={loadingText}>{children}</ButtonSpinner>
          {children && !loadingText && (
            <nature.span className='opacity-0'>{children}</nature.span>
          )}
        </>
      ) : (
        children
      )}
      {rightIcon && !isLoading && (
        <ButtonIcon
          css={{
            marginLeft: iconSpacing,
          }}
        >
          {rightIcon}
        </ButtonIcon>
      )}
    </nature.button>
  );
});

if (__DEV__) {
  Button.displayName = 'Button';
  ButtonSpinner.displayName = 'ButtonSpinner';
}
