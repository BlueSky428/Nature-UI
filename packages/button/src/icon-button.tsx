import { clsx, forwardRef } from '@nature-ui/system';
import { __DEV__ } from '@nature-ui/utils';
import * as React from 'react';
import { Button, ButtonProps } from './button';

type Omitted =
  | 'leftIcon'
  | 'rightIcon'
  | 'loadingText'
  | 'iconSpacing'
  | 'isIconButton';

interface BaseButtonProps extends Omit<ButtonProps, Omitted> {}

export interface IconButtonProps extends BaseButtonProps {
  /**
   * The icon to be used in the button.
   * @type React.ReactElement
   */
  icon?: React.ReactElement;
  /**
   * If `true`, the button will be perfectly round. Else, it'll be slightly round
   */
  isRound?: boolean;
  /**
   * A11y: A label that describes the button
   */
  'aria-label'?: string;
}

export const IconButton = forwardRef<IconButtonProps, 'button'>(
  (props, ref) => {
    const {
      icon,
      children,
      isRound,
      'aria-label': ariaLabel,
      className = '',
      color = 'gray-200',
      as, // FIXME: if this is included it breaks the code
      ...rest
    } = props;

    /**
     * Passing the icon as prop or children should work
     */
    const element = icon || children;
    const _children = React.isValidElement(element)
      ? React.cloneElement(element, {
          // @ts-ignore
          'aria-hidden': true,
          focusable: false,
        })
      : null;
    const _className = clsx(className, {
      'rounded-full': isRound,
      'rounded-md': !isRound,
    });

    return (
      <Button
        ref={ref}
        color={color}
        aria-label={ariaLabel}
        className={_className}
        isIconButton
        {...rest}
      >
        {_children}
      </Button>
    );
  },
);

if (__DEV__) {
  IconButton.displayName = 'IconButton';
}
