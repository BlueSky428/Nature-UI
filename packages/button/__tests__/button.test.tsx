import * as React from 'react';
import { render } from '@nature-ui/test-utils';

import { Button } from '../src';

describe('@nature-ui/button', () => {
  test('Button rernders correctly', () => {
    const { asFragment } = render(<Button />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('shows spinner if isLoading', () => {
    const { asFragment, getByText } = render(<Button isLoading>Email</Button>);

    expect(asFragment()).toMatchSnapshot();

    // "Loading..." visually hidden label shown
    expect(getByText('Loading...')).toBeInTheDocument();
  });

  test('shows spinner and loading text if isLoading and loadingText', () => {
    const { asFragment, queryByText, getByText } = render(
      <Button isLoading loadingText='Submitting'>
        Submit
      </Button>
    );

    expect(asFragment()).toMatchSnapshot();

    // children text is replaced by `loadingText`
    getByText('Submitting');
    expect(queryByText('Submit')).toBeNull();
  });

  test('has the proper aria attributes', () => {
    const { rerender, getByRole } = render(<Button>Hello</Button>);

    // button has role="button"
    const button = getByRole('button');

    expect(button).not.toHaveAttribute('aria-disabled');

    // isLoading sets aria-disabled="true"
    rerender(<Button isLoading>Hello</Button>);
    expect(button).toHaveAttribute('data-loading', '');

    // isDisabled sets aria-disabled="true"
    rerender(<Button isDisabled>Hello</Button>);
    expect(button).toHaveAttribute('disabled', '');
  });
});
