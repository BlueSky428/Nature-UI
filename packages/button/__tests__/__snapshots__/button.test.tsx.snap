// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@nature-ui/button Button rernders correctly 1`] = `
<DocumentFragment>
  <button
    class="css-mqia37 focus:shadow-outline focus:outline-none rounded font-semibold relative overflow-hidden align-middle inline-flex justify-center items-center leading-normal css-1x86u9b bg-gray-200 text-gray-700 hover:bg-gray-300 px-4 css-15pri0z-Button"
  />
</DocumentFragment>
`;

exports[`@nature-ui/button shows spinner and loading text if isLoading and loadingText 1`] = `
<DocumentFragment>
  <button
    class="css-mqia37 focus:shadow-outline focus:outline-none rounded font-semibold relative overflow-hidden align-middle inline-flex justify-center items-center leading-normal css-1x86u9b bg-gray-200 text-gray-700 hover:bg-gray-300 opacity-50 cursor-not-allowed px-4 css-15pri0z-Button"
    data-loading=""
    disabled=""
  >
    <span
      class="align-middle relative"
    >
      <span
        class="css-hxvqwd animation-1nvx475 inline-block overflow-hidden border-2 border-transparent border-t-2 align-middle w-4 h-4 text-currentColor"
        style="border-top: 2px solid currentColor; border-left: 2px solid currentColor;"
      >
        <span
          class="css-iaws77-VisuallyHidden"
        >
          Loading...
        </span>
      </span>
      <span
        class="ml-2"
      >
        Submitting
      </span>
    </span>
  </button>
</DocumentFragment>
`;

exports[`@nature-ui/button shows spinner if isLoading 1`] = `
<DocumentFragment>
  <button
    class="css-mqia37 focus:shadow-outline focus:outline-none rounded font-semibold relative overflow-hidden align-middle inline-flex justify-center items-center leading-normal css-1x86u9b bg-gray-200 text-gray-700 hover:bg-gray-300 opacity-50 cursor-not-allowed px-4 css-15pri0z-Button"
    data-loading=""
    disabled=""
  >
    <span
      class="align-middle absolute"
    >
      <span
        class="css-hxvqwd animation-1nvx475 inline-block overflow-hidden border-2 border-transparent border-t-2 align-middle w-4 h-4 text-currentColor"
        style="border-top: 2px solid currentColor; border-left: 2px solid currentColor;"
      >
        <span
          class="css-iaws77-VisuallyHidden"
        >
          Loading...
        </span>
      </span>
    </span>
    <span
      class="opacity-0"
    >
      Email
    </span>
  </button>
</DocumentFragment>
`;
